ospf6d/ospf6_asbr.c

```diff
@@ -1505,6 +1540,145 @@ static void ospf6_redistribute_show_config(struct vty *vty, struct ospf6 *ospf6,
 		vty_out(vty, "Total %d routes\n", total);
 }
 
+static void ospf6_redistribute_default_set(struct ospf6 *ospf6, int originate)
+{
+	struct prefix_ipv6 p;
+	struct in6_addr nexthop;
+	int cur_originate = ospf6->default_originate;
+
+	p.family = AF_INET6;
+	p.prefixlen = 0;
+
+	memset(&p.prefix, 0, sizeof(p.prefix));
+	memset(&nexthop, 0, sizeof(nexthop));
+
+	ospf6->default_originate = originate;
+
+	switch (cur_originate) {
+	case DEFAULT_ORIGINATE_NONE:
+		break;
+	case DEFAULT_ORIGINATE_ZEBRA:
+		zclient_redistribute_default(ZEBRA_REDISTRIBUTE_DEFAULT_DELETE,
+					     zclient, AFI_IP6, ospf6->vrf_id);
+		ospf6_asbr_redistribute_remove(DEFAULT_ROUTE, 0,
+					       (struct prefix *)&p, ospf6);
+
+		break;
+	case DEFAULT_ORIGINATE_ALWAYS:
+		ospf6_asbr_redistribute_remove(DEFAULT_ROUTE, 0,
+					       (struct prefix *)&p, ospf6);
+		break;
+	}
+
+	switch (originate) {
+	case DEFAULT_ORIGINATE_NONE:
+		break;
+	case DEFAULT_ORIGINATE_ZEBRA:
+		zclient_redistribute_default(ZEBRA_REDISTRIBUTE_DEFAULT_ADD,
+					     zclient, AFI_IP6, ospf6->vrf_id);
+
+		break;
+	case DEFAULT_ORIGINATE_ALWAYS:
+		ospf6_asbr_redistribute_add(DEFAULT_ROUTE, 0,
+					    (struct prefix *)&p, 0, &nexthop, 0,
+					    ospf6);
+		break;
+	}
+}
+
+/* Default Route originate. */
+DEFPY (ospf6_default_route_originate,
+       ospf6_default_route_originate_cmd,
+       "default-route originate [{always|metric (0-16777214)$mval|metric-type (1-2)$mtype|route-map WORD$rtmap}]",
+       "Control distribution of default route\n"
+       "Distribute a default route\n"
+       "Always advertise default route\n"
+       "OSPFv3 default metric\n"
+       "OSPFv3 metric\n"
+       "OSPFv3 metric type for default routes\n"
+       "Set OSPFv3 External Type 1/2 metrics\n"
+       "Route map reference\n"
+       "Pointer to route-map entries\n")
+{
+	int default_originate = DEFAULT_ORIGINATE_ZEBRA;
+	int idx = 0;
+	struct ospf6_redist *red;
+	bool sameRtmap = false;
+
+	VTY_DECLVAR_CONTEXT(ospf6, ospf6);
+
+	int cur_originate = ospf6->default_originate;
+
+	OSPF6_CMD_CHECK_RUNNING(ospf6);
+
+	red = ospf6_redist_add(ospf6, DEFAULT_ROUTE, 0);
+
+	if (argv_find(argv, argc, "always", &idx))
+		default_originate = DEFAULT_ORIGINATE_ALWAYS;
+
+	idx = 0;
+	if (!argv_find(argv, argc, "metric", &idx))
```

Comment:
I don't think this is necessary, wouldn't it be easier to just check for `mval_str == NULL`?

Same for `metric-type` (`mtype`) and `always`.

Reply:
Yes, but the only problem I faced was mval which should be a long data-type gets 0 as the default value if the metric parameter is not used. But mval can also have 0 as valid input. To fix this conflict I have used this way of setting the value when the metric parameter is not used. For metric-type and always I have made the fix and use the NULL comparison. Is there a better way to handle the conflict?

Reply:
Integer, addresses and other types they also have the corresponding `<var>_str` version passed to `DEFPY`.

If you want to check for missing values you can just do, for an example:

```c
if (mval_str == NULL)
  mval = -1;
```

Reply:
Ok. Thanks. I have fixed it now.